[Collatz Conjecture Dig Deeper]: Corrections & Edit (#4250)
exercism/python: exercises/practice/collatz-conjecture/.approaches/introduction.md

# Introduction

> There are multiple approaches to solving the Collatz Conjecture exercise idiomatically in Python.

Reviewer: This line is a little awkward. I think moving "idiomatic" would help:

```suggestion
There are multiple idiomatic approaches to solving the Collatz Conjecture exercise in Python.
```

You can, for example, use a `while loop` or a `recursive` function.
You can also solve the exercise by using `if`/`else` statements or the `ternary operator`.

## General guidance

The key to this exercise is to check if the number is even or odd and then perform the correct operation.

> Under this process (_if the data is not [pathological][collatz-pathological]_), the result will settle at 1.

Reviewer: This link does seem useful, but it being right here confused me, as it does not mention the word "pathological".

Author: I think we keep it.  Unless you'd like to replace it with a def of "pathological data"?  The point being ... and I was frustrated I couldn't find a clearer article, but if you use something like 27 (which takes 111 steps to settle), or some other numbers, it can take a long time to come to 1.  That being said, we know all numbers up to about 2^71 **will** settle, it is just a matter of the number of steps.

Author: Open to suggestions on how to describe the link.

Author: went with

> (_if the input number doesn't create an excessively [long cycle][collatz-pathological]_), the result will settle at 1.

    return counter
```

For more information, read the [Ternary operator approach][approach-ternary-operator].

## Approach: Recursive function

In this approach we use a `recursive function`.
A recursive function is a function that calls itself.
This approach can be more concise than other approaches, but may or may not be more readable for some audiences.

You might not want to use this approach due to Python's [`recursion` limit][recursion-limit], which has a default of 1000 stack frames.
While the current tests for this exercise all have input that is easily calculated in under the `recursion` limit, this is not true for arbitrary numbers.

> Numbers with a large number of steps may need to employ memoization or other techniques to make `recursion` work.

Reviewer: This sentence is confusing, maybe it could be something like this instead:

```suggestion
To make this approach work with large numbers of steps, techniques such as memoization may need to be employed.
```
